sql/src/main/java/io/crate/analyze/AnalyzedTableElements.java

```diff
@@ -158,6 +181,76 @@ public void finalizeAndValidate() {
         validatePrimaryKeys();
     }
 
+    private void validateGeneratedColumns(TableIdent tableIdent,
+                                          AnalysisMetaData analysisMetaData,
+                                          ParameterContext parameterContext) {
+        List<ReferenceInfo> tableReferenceInfos = new ArrayList<>();
+        for (AnalyzedColumnDefinition columnDefinition : columns) {
+            buildReferenceInfo(tableIdent, columnDefinition, tableReferenceInfos);
+        }
+        try {
+            TableInfo tableInfo = analysisMetaData.referenceInfos().getTableInfo(tableIdent);
+            tableReferenceInfos.addAll(tableInfo.columns());
+        } catch (Exception e) {
+            // table does not exists yet, no more references to add
+        }
+
+        ExpressionAnalyzer expressionAnalyzer = new ExpressionReferenceAnalyzer(
+                analysisMetaData, parameterContext, tableReferenceInfos);
+        SymbolFormatter formatter = new SymbolFormatter(analysisMetaData.functions());
+        for (AnalyzedColumnDefinition columnDefinition : columns) {
+            if (columnDefinition.generatedExpression() != null) {
+                processGeneratedExpression(expressionAnalyzer, formatter, columnDefinition);
+            }
+        }
+    }
+
+    private void processGeneratedExpression(ExpressionAnalyzer expressionAnalyzer,
+                                            SymbolFormatter symbolFormatter,
+                                            AnalyzedColumnDefinition columnDefinition) {
+        // validate expression
+        Symbol function = expressionAnalyzer.convert(columnDefinition.generatedExpression(), new ExpressionAnalysisContext());
+
+        String formattedExpression;
+        DataType valueType = function.valueType();
+        DataType definedType = columnDefinition.dataType() == null ? null : DataTypes.ofMappingNameSafe(columnDefinition.dataType());
+
+        // check for optional defined type and add `cast` to expression if possible
+        if (definedType != null && !definedType.equals(valueType)) {
+            Preconditions.checkArgument(valueType.isConvertableTo(definedType),
+                    "generated expression value type '%s' not supported for conversion to '%s'", valueType, definedType.getName());
+
+            Function castFunction = new Function(CastFunctionResolver.functionInfo(valueType, definedType, false), Collections.singletonList(function));
```

Comment:
Collections.singletonList isn't mutable but the arguments to a Function symbol should be in a mutable list. The Function symbol has a `setArgument` function which would break otherwise.